parser.js

```diff
@@ -94,81 +117,69 @@ const metadataRules = {
     rules: [
-      ['meta[name="keywords"]', node => node.element.getAttribute('content')],
+      ['meta[name="keywords"]', element => element.getAttribute('content')],
     ],
     processors: [
-      (keywords) => keywords.split(',').map((keyword) => keyword.trim()),
+      (keywords, context) => keywords.split(',').map((keyword) => keyword.trim())
     ]
   },
 
   title: {
     rules: [
-      ['meta[property="og:title"]', node => node.element.getAttribute('content')],
-      ['meta[name="twitter:title"]', node => node.element.getAttribute('content')],
-      ['meta[property="twitter:title"]', node => node.element.getAttribute('content')],
-      ['meta[name="hdl"]', node => node.element.getAttribute('content')],
-      ['title', node => node.element.text],
+      ['meta[property="og:title"]', element => element.getAttribute('content')],
+      ['meta[name="twitter:title"]', element => element.getAttribute('content')],
+      ['meta[property="twitter:title"]', element => element.getAttribute('content')],
+      ['meta[name="hdl"]', element => element.getAttribute('content')],
+      ['title', element => element.text],
     ],
   },
 
   type: {
     rules: [
-      ['meta[property="og:type"]', node => node.element.getAttribute('content')],
+      ['meta[property="og:type"]', element => element.getAttribute('content')],
     ],
   },
 
   url: {
     rules: [
-      ['meta[property="og:url"]', node => node.element.getAttribute('content')],
-      ['link[rel="canonical"]', node => node.element.getAttribute('href')],
+      ['meta[property="og:url"]', element => element.getAttribute('content')],
+      ['link[rel="canonical"]', element => element.getAttribute('href')],
     ],
+    defaultValue: (context) => context.url,
     processors: [
       (url, context) => makeUrlAbsolute(context.url, url)
     ]
   },
 
   provider: {
     rules: [
-      ['meta[property="og:site_name"]', node => node.element.getAttribute('content')]
-    ]
+      ['meta[property="og:site_name"]', element => element.getAttribute('content')]
+    ],
+    defaultValue: (context) => getProvider(parseUrl(context.url))
   },
 };
 
-function getMetadata(doc, url, rules) {
+function getMetadata(doc, url, customRuleSets) {
   const metadata = {};
-  const context = {url};
-  const ruleSet = rules || metadataRules;
-
-  Object.keys(ruleSet).map(metadataKey => {
-    const metadataRule = ruleSet[metadataKey];
-
-    if(Array.isArray(metadataRule.rules)) {
-      const builtRule = buildRuleset(metadataKey, metadataRule.rules, metadataRule.processors);
-      metadata[metadataKey] = builtRule(doc, context);
-    } else {
-      metadata[metadataKey] = getMetadata(doc, url, metadataRule);
-    }
-  });
+  const context = {
+    url,
+  };
 
-  if(!metadata.url) {
-    metadata.url = url;
-  }
+  const ruleSets = customRuleSets || metadataRuleSets;
 
-  if(url && !metadata.provider) {
-    metadata.provider = getProvider(url);
-  }
+  Object.keys(ruleSets).map(ruleSetKey => {
```

Comment:
It might be slightly more efficient and readable to do:

```
for (const ruleSetKey in ruleSets) {
  const ruleSet = ruleSets[ruleSetKey];
  const builtRuleSet = buildRuleSet(ruleSet);

  metadata[ruleSetKey] = builtRuleSet(doc, context);
}
```